README.md
# Python Port Scanner v0.3

An easy to use Python3 package that could perform port scanning conveniently.

An output example is showed as following:  
![Output Example](https://github.com/YaokaiYang-assaultmaster/py3PortScanner/blob/master/ExampleGraph/portscanner_output_new.png)

## Installation  
1. Clone or download this repository.
2. Install the package using `python setup.py install`.   
3. Voilà! You are ready to go!

## QuickStart
1. Add `from pyportscanner import pyscanner` to the beginning of your code.
2. Initialize a new PortScanner object using `scanner = pyscanner.PortScanner(target_ports=100, timeout=10, verbose=True)`.
3. Then call `scanner.scan(objective)` to perform a port scan to a specific target.
The target could either be an IPv4 address or a host name.
4. __Note that the total scan time for a target website is highly related to the timeout value set for the Scanner object. Thus for the seek of efficiency, the timeout should not be too long.__

## Documentation 
### _class pyportscanner.pyscanner.PortScanner(target_ports=None, thread_limit=100, timeout=10, verbose=False)_
PortScanner is the class provides methods to execute the port scan request. A PortScanner object is needed for performing
the port scan request.  

- _target_ports_ can be a list or int. If this args is a list, then the list of ports specified by it is going to be scanned, 
default to all ports we have in file. If this args is an int, then it specifies the top X number of ports to be scanned based on usage
frequency rank.
- _thread_limit_ is the number of thread being used for scan.  
- _timeout_ is the timeout for the socket to wait for a response.
- _verbose_ specifies whether the results would be print out or not. If `True`, results will be print out. 

### _Functions_  
__PortScanner.scan(objective, message = '')__ 

Scan an objective with the given message included in the packets sent out.  

- _objective_ is the target that is going to be scanned. It could be an IPv4 address or a hostname.  
- _message_ is the message that is going to be included in the scanning packets sent out. If not provided, no message will be included in the packets.    

An example usage case is showed in [_examples/PortScanExample.py_](https://github.com/YaokaiYang-assaultmaster/py3PortScanner/blob/master/examples/PortScanExample.py).  

## [Change logs](https://github.com/YaokaiYang-assaultmaster/py3PortScanner/blob/master/CHANGELOG.md)


CHANGELOG.md
# Changelog

Versions are monotonically increased based on Semantic Versioning.

***

## V0.3 (10/09/2018)

### Backward incompatible changes:
No longer compatible with python 2.7.
For a version compatible with python 2.7, refer to [v0.2](https://github.com/YaokaiYang-assaultmaster/PythonPortScanner).

### Deprecations
None

### Changes
Migrate to python version >= 3.0.
Integrate the whole [nmap](https://github.com/nmap/nmap) port list into the project.

## V0.2 (09/12/2017)

### Backward incompatible changes
None

### Deprecations
None

### Changes
Fixed [issue #1](https://github.com/YaokaiYang-assaultmaster/PythonPortScanner/issues/1). 
Specifically, fixed Windows has no `SO_REUSEPORT` for TCP socket error and change message 
encoding to `utf-8` before sending out. 

***

## V0.1 (05/30/2017)

### Backward incompatible changes
None

### Deprecations
None

### Changes
Initialized the whole project. First version finished. 


examples/PortScanExample.py
from pyportscanner import pyscanner


def main():
    # Initialize a Scanner object that will scan top 50 commonly used ports.
    scanner = pyscanner.PortScanner(target_ports=100, verbose=True)

    host_name = 'google.com'

    message = 'put whatever message you want here'

    '''
    output contains a dictionary of {port:status} pairs
    in which port is the list of ports we scanned 
    and status is either 'OPEN' or 'CLOSE'
    '''

    res = scanner.scan(host_name, message)
    '''
    ************************************************************
    
    Start scanning website: google.com
    Server ip is: 172.217.6.110
    80/TCP    :       OPEN
    
    2000/UDP  :       OPEN
    
    5060/UDP  :       OPEN
    
    8008/TCP  :       OPEN
    
    Host google.com scanned in  10.033627033233643 seconds
    Scan completed!
    '''

    '''
    > res
    {9: 'CLOSE',
     17: 'CLOSE',
     19: 'CLOSE',
     21: 'CLOSE',
     22: 'CLOSE',
     25: 'CLOSE',
     26: 'CLOSE',
     49: 'CLOSE',
     53: 'CLOSE',
     67: 'CLOSE',
     68: 'CLOSE',
     69: 'CLOSE',
     80: 'OPEN',
     81: 'CLOSE',
     88: 'CLOSE',
     110: 'CLOSE',
     111: 'CLOSE',
     113: 'CLOSE',
     123: 'CLOSE',
     135: 'CLOSE',
     136: 'CLOSE',
     137: 'CLOSE',
     138: 'CLOSE',
     139: 'CLOSE',
     143: 'CLOSE',
     158: 'CLOSE',
     161: 'CLOSE',
     162: 'CLOSE',
     177: 'CLOSE',
     179: 'CLOSE',
     199: 'CLOSE',
     427: 'CLOSE',
     443: 'CLOSE',
     445: 'CLOSE',
     465: 'CLOSE',
     514: 'CLOSE',
     515: 'CLOSE',
     518: 'CLOSE',
     520: 'CLOSE',
     548: 'CLOSE',
     554: 'CLOSE',
     587: 'CLOSE',
     593: 'CLOSE',
     623: 'CLOSE',
     626: 'CLOSE',
     631: 'CLOSE',
     646: 'CLOSE',
     993: 'CLOSE',
     995: 'CLOSE',
     999: 'CLOSE',
     1022: 'CLOSE',
     1025: 'CLOSE',
     1026: 'CLOSE',
     1027: 'CLOSE',
     1029: 'CLOSE',
     1030: 'CLOSE',
     1031: 'CLOSE',
     1032: 'CLOSE',
     1433: 'CLOSE',
     1434: 'CLOSE',
     1645: 'CLOSE',
     1646: 'CLOSE',
     1718: 'CLOSE',
     1719: 'CLOSE',
     1720: 'CLOSE',
     1723: 'CLOSE',
     1812: 'CLOSE',
     1813: 'CLOSE',
     1900: 'CLOSE',
     2000: 'OPEN',
     2001: 'CLOSE',
     2049: 'CLOSE',
     2222: 'CLOSE',
     2223: 'CLOSE',
     3283: 'CLOSE',
     3389: 'CLOSE',
     3456: 'CLOSE',
     3703: 'CLOSE',
     4045: 'CLOSE',
     4500: 'CLOSE',
     5000: 'CLOSE',
     5060: 'OPEN',
     5353: 'CLOSE',
     5355: 'CLOSE',
     5666: 'CLOSE',
     5900: 'CLOSE',
     8000: 'CLOSE',
     8008: 'OPEN',
     8080: 'CLOSE',
     8443: 'CLOSE',
     8888: 'CLOSE',
     9200: 'CLOSE',
     10000: 'CLOSE',
     17185: 'CLOSE',
     20031: 'CLOSE',
     31337: 'CLOSE',
     32768: 'CLOSE',
     32769: 'CLOSE',
     32770: 'CLOSE',
     32771: 'CLOSE'}
    '''


if __name__ == "__main__":
    main()


pyportscanner/etc/helper.py
import re
from urllib.parse import urlparse
import pkg_resources

from pyportscanner.etc.service_port import ServicePort


def read_input():
    """
    Read the 'nmap-services.txt' file and store all the information into
    a dict() of {port, ServicePort} pairs for reference later.
    """
    resource_package = __name__
    resource_path = 'nmap-services.dat'
    resource = pkg_resources.resource_stream(resource_package, resource_path)

    port_map = dict()
    line_regex = '([a-zA-Z0-9-]+)\s+(\d+)/(\w+)\s+(\d+\.\d+)\s+(\#.*)'
    pattern = re.compile(line_regex)
    for line in resource:
        line = line.decode('utf-8')
        # skip comments
        if line.startswith('#'):
            continue
        result = pattern.match(line)
        if result:
            service_name = result.group(1)
            port_num = int(result.group(2))
            proto = result.group(3)
            freq = float(result.group(4))
            service_port = ServicePort(service_name, port_num, proto, freq)
            if port_num not in port_map:
                port_map[port_num] = service_port
            elif port_map[port_num].freq < freq:
                # only keeps the port and protocol with highest usage frequency
                port_map[port_num] = service_port

    return port_map


def get_domain(url):
    """
    Return the hostname, or domain name, of a url.
    e.g. for 'http://google.com/path', it will return google.com
    :param url: String url
    :return: hostname of the url (note the hostname does not include the protocol part)
    """
    if not url:
        return u""
    full_url = u"http://{}"
    if not url.startswith(u"http://") and not url.startswith(u"https://"):
        # otherwise the urlparse will return empty values
        full_url = full_url.format(url)
    else:
        full_url = url
    parse_result = urlparse(full_url)
    return parse_result.hostname


pyportscanner/PortScanner.py
# -*- coding: utf-8 -*-

import platform
import socket
import threading
import time

import concurrent.futures

from pyportscanner.etc.helper import read_input, get_domain


class PortScanner:
    @classmethod
    def __usage(cls):
        """
        Return the usage information for invalid input host name.
        """
        print('Python Port Scanner')
        print('Please make sure the input host name is in the form of "foo.com" or "http://foo.com!"\n')

    @property
    def timeout_val(self):
        return self.__timeout

    @timeout_val.setter
    def timeout_val(self, timeout):
        if timeout != int(timeout):
            raise TypeError('Timeout must be an integer')
        elif timeout <= 0:
            raise ValueError(
                'Invalid timeout value: {}.'
                'Timeout must be greater than 0'.format(timeout)
            )
        else:
            self.__timeout = timeout

    @timeout_val.getter
    def timeout_val(self):
        return self.__timeout

    @property
    def thread_limit(self):
        return self.__thread_limit

    @thread_limit.setter
    def thread_limit(self, thread_limit):
        if thread_limit != int(thread_limit):
            raise TypeError('thread limit must be an integer')
        elif thread_limit <= 0 or thread_limit > 50000:
            self.__thread_limit = 100
            raise ValueError(
                'Invalid thread limit {}.'
                'Thread limit must be within 0 to 50000 '.format(thread_limit)
            )
        else:
            self.__thread_limit = thread_limit

    @thread_limit.getter
    def thread_limit(self):
        return self.__thread_limit

    def __init__(self, target_ports=None, thread_limit=100, verbose=False):
        """
        Constructor of a PortScanner object. If target_ports is a list, this list of ports will be used as
        the port list to be scanned. If the target_ports is a int, it should be 50, 100 or 1000, indicating
        which default list will be used.

        :param target_ports: If this args is a list, then this list of ports that is going to be scanned,
        default to all ports we have in file.
        If this args is an int, then it specifies the top X number of ports to be scanned based on usage
        frequency rank.
        :type target_ports: list or int
        :param verbose: If True, the scanner will print out scanning result. If False, the scanner
        will scan silently.
        :type verbose boolean
        """
        # default ports to be scanned are all ports in file
        self.__port_map = read_input()

        # default thread number limit
        self.__thread_limit = thread_limit

        # default connection timeout time in seconds
        self.__timeout = 10

        self.__verbose = verbose

        if target_ports is None:
            self.targets = self.__port_map.keys()
        elif type(target_ports) == list:
            self.targets = target_ports
        elif type(target_ports) == int:
            self.targets = self.extract_list(target_ports)

    def extract_list(self, target_port_rank):
        """
        Extract the top X ranked ports based usage frequency.
        If a number greater than the total number of ports we have is specified, scan all ports.

        :param target_port_rank: top X commonly used port list to be returned.
        :return: top X commonly used port list.
        """
        if target_port_rank <= 0:
            raise ValueError(
                'Invalid input {}. No ports can be selected'.format(target_port_rank)
            )

        service_port_list = sorted(self.__port_map.values())
        port_list = list(ele.port_num for ele in service_port_list)
        return sorted(port_list[:target_port_rank])

    def get_target_ports(self):
        """
        Return the list of ports being scanned.

        :return: list of ports scanned by current Scanner object.
        :rtype: list
        """
        return self.targets

    def get_top_k_ports(self, k):
        """
        Return top K commonly used ports.

        :param k: top K ports to be returned.
        :type k: int
        :return: top K commonly used ports.
        :rtype: list
        """
        port_list = self.extract_list(k)
        return port_list

    def scan(self, host_name, message=''):
        """
        This is the function need to be called to perform port scanning.

        :param host_name: the hostname that is going to be scanned
        :param message: the message that is going to be included in the scanning packets
        in order to prevent ethical problem (default: '').
        :return: a dict object containing the scan results for a given host in the form of
        {port_number: status}
        :rtype: dict
        """
        host_name = get_domain(host_name)

        if self.__verbose:
            print('\n')
            print('*' * 60 + '\n')
            print('Start scanning website: {}'.format(host_name))

        try:
            server_ip = socket.gethostbyname(host_name)
            if self.__verbose:
                print('Server ip is: {}'.format(str(server_ip)))

        except socket.error:
            # If the DNS resolution of a website cannot be finished, abort that website.
            if self.__verbose:
                print('Hostname {} unknown! Scan failed'.format(host_name))
                self.__usage()
            return {}

        start_time = time.time()
        output = self.__scan_ports(server_ip, message)
        stop_time = time.time()

        if self.__verbose:
            print('Host {} scanned in  {} seconds'.format(host_name, stop_time - start_time))
            print('Scan completed!\n')

        return output

    def __scan_ports(self, ip, message):
        """
        Controller of the __scan_ports_helper() function

        :param ip: the ip address that is being scanned
        :type ip: str
        :param delay: the time in seconds that a TCP socket waits until timeout
        :type delay: int
        :param message: the message that is going to be included in the scanning packets,
        in order to prevent ethical problem, default to ''.
        :type message: str
        :return: a dict that stores result in {port, status} style pairs.
        status can be 'OPEN' or 'CLOSE'.
        """
        output = dict()
        for port in self.targets:
            output[port] = 'CLOSE'
        
        futures = list()
        with concurrent.futures.ThreadPoolExecutor(max_workers=self.__thread_limit) as executor:
            for port in self.targets:
                future = executor.submit(self.__TCP_connect, ip, port, message)
                futures.append(future)
                while len(futures) >= self.__thread_limit:
                    self.__check_futures(output, futures)
                    time.sleep(0.01)

        # Print opening ports from small to large
        if self.__verbose:
            for port in self.targets:
                if output[port] == 'OPEN':
                    service = self.__port_map.get(port, None)
                    if service:
                        port_proto = '{}/{}'.format(port, service.proto.upper())
                    else:
                        port_proto = '{}/{}'.format(port, 'UNKNOWN')
                    print('{:10}: {:>10}\n'.format(port_proto, output[port]))

        return output

    @classmethod
    def __check_futures(cls, output, futures):
        """
        Check the executing status of Futures and retrieve the results from them.
        :param output: dict for storing the results
        :param futures: list of concurrent.futures.Future object
        """
        for future in futures:
            if future.done():
                try:
                    port, status = future.result()
                    output[port] = status
                except Exception:
                    pass
                futures.remove(future)

    def __TCP_connect(self, ip, port_number, message):
        """
        Perform status checking for a given port on a given ip address using TCP handshake

        :param ip: the ip address that is being scanned
        :type ip: str
        :param port_number: the port that is going to be checked
        :type port_number: int
        :param message: the message that is going to be included in the scanning packets,
        in order to prevent ethical problem, default to ''.
        :type message: str
        """
        # Initialize the TCP socket object based on different operating systems.
        # All systems except for 'Windows' will be treated equally.
        curr_os = platform.system()
        if curr_os == 'Windows':
            TCP_sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
            TCP_sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
            TCP_sock.settimeout(self.__timeout)
        else:
            TCP_sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
            TCP_sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEPORT, 1)
            TCP_sock.settimeout(self.__timeout)

        b_message = message.encode('utf-8', errors='replace')
        # Initialize a UDP socket to send scanning alert message if there exists an non-empty message
        if message:
            UDP_sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
            UDP_sock.sendto(b_message, (ip, int(port_number)))

        try:
            result = TCP_sock.connect_ex((ip, int(port_number)))
            if message != '':
                TCP_sock.sendall(b_message)

            TCP_sock.close()
            # If the TCP handshake is successful, the port is OPEN. Otherwise it is CLOSE
            if result == 0:
                return port_number, 'OPEN'
            else:
                return port_number, 'CLOSE'

        except socket.error as e:
            # Failed to perform a TCP handshake means the port is probably close.
            return port_number, 'CLOSE'


pyportscanner/pyscanner.py
import concurrent.futures
import platform
import socket
import time
from collections import deque

from pyportscanner.etc.helper import read_input, get_domain


class PortScanner:
    @classmethod
    def __usage(cls):
        """
        Return the usage information for invalid input host name.
        """
        print('Python Port Scanner')
        print('Please make sure the input host name is in the form of "foo.com" or "http://foo.com!"\n')

    @property
    def timeout_val(self):
        return self.__timeout

    @timeout_val.setter
    def timeout_val(self, timeout):
        if timeout != int(timeout):
            raise TypeError('Timeout must be an integer')
        elif timeout <= 0:
            raise ValueError(
                'Invalid timeout value: {}.'
                'Timeout must be greater than 0'.format(timeout)
            )
        else:
            self.__timeout = timeout

    @timeout_val.getter
    def timeout_val(self):
        return self.__timeout

    @property
    def thread_limit(self):
        return self.__thread_limit

    @thread_limit.setter
    def thread_limit(self, thread_limit):
        if thread_limit != int(thread_limit):
            raise TypeError('thread limit must be an integer')
        elif thread_limit <= 0 or thread_limit > 50000:
            self.__thread_limit = 100
            raise ValueError(
                'Invalid thread limit {}.'
                'Thread limit must be within 0 to 50000 '.format(thread_limit)
            )
        else:
            self.__thread_limit = thread_limit

    @thread_limit.getter
    def thread_limit(self):
        return self.__thread_limit

    def __init__(self, target_ports=None, thread_limit=100, timeout=10, verbose=False):
        """
        Constructor of a PortScanner object. If target_ports is a list, this list of ports will be used as
        the port list to be scanned. If the target_ports is a int, it should be 50, 100 or 1000, indicating
        which default list will be used.

        :param target_ports: If this args is a list, then this list of ports is going to be scanned,
        default to all ports we have in file.
        If this args is an int, then it specifies the top X number of ports to be scanned based on usage
        frequency rank.
        :type target_ports: list or int
        :param verbose: If True, the scanner will print out scanning result. If False, the scanner
        will scan silently.
        :type verbose boolean
        """
        # default ports to be scanned are all ports in file
        self.__port_map = read_input()

        # default thread number limit
        self.__thread_limit = thread_limit

        # default connection timeout time in seconds
        self.__timeout = timeout

        self.__verbose = verbose

        if target_ports is None:
            self.targets = self.__port_map.keys()
        elif type(target_ports) == list:
            self.targets = target_ports
        elif type(target_ports) == int:
            self.targets = self.extract_list(target_ports)

    def extract_list(self, target_port_rank):
        """
        Extract the top X ranked ports based usage frequency.
        If a number greater than the total number of ports we have is specified, scan all ports.

        :param target_port_rank: top X commonly used port list to be returned.
        :return: top X commonly used port list.
        """
        if target_port_rank <= 0:
            raise ValueError(
                'Invalid input {}. No ports can be selected'.format(target_port_rank)
            )

        service_port_list = sorted(self.__port_map.values())
        port_list = list(ele.port_num for ele in service_port_list)
        return sorted(port_list[:target_port_rank])

    def get_target_ports(self):
        """
        Return the list of ports being scanned.

        :return: list of ports scanned by current Scanner object.
        :rtype: list
        """
        return self.targets

    def get_top_k_ports(self, k):
        """
        Return top K commonly used ports.

        :param k: top K ports to be returned.
        :type k: int
        :return: top K commonly used ports.
        :rtype: list
        """
        port_list = self.extract_list(k)
        return port_list

    def scan(self, objective, message=''):
        """
        This is the function need to be called to perform port scanning.

        :param objective: the objective that is going to be scanned. Could be a host name or an IPv4 address.
        :param message: the message that is going to be included in the scanning packets
        in order to prevent ethical problem (default: '').
        :return: a dict object containing the scan results for a given host in the form of
        {port_number: status}
        :rtype: dict
        """
        try:
            socket.inet_aton(objective)
            host_name = objective
        except OSError or socket.error:
            # this is not an valid IPv4 address
            host_name = get_domain(objective)

        if self.__verbose:
            print('\n')
            print('*' * 60 + '\n')
            print('Start scanning target: {}'.format(host_name))

        try:
            server_ip = socket.gethostbyname(host_name)
            if self.__verbose:
                print('Target IP is: {}'.format(str(server_ip)))

        except socket.error:
            # If the DNS resolution of a website cannot be finished, abort the host.
            if self.__verbose:
                print('Target {} unknown! Scan failed.'.format(host_name))
                self.__usage()
            return {}

        start_time = time.time()
        output = self.__scan_ports(server_ip, message)
        stop_time = time.time()

        if self.__verbose:
            print('Target {} scanned in  {} seconds'.format(host_name, stop_time - start_time))
            print('Scan completed!\n')

        return output

    def __scan_ports(self, ip, message):
        """
        Controller of the __scan_ports_helper() function

        :param ip: the ip address that is being scanned
        :type ip: str
        :param delay: the time in seconds that a TCP socket waits until timeout
        :type delay: int
        :param message: the message that is going to be included in the scanning packets,
        in order to prevent ethical problem, default to ''.
        :type message: str
        :return: a dict that stores result in {port, status} style pairs.
        status can be 'OPEN' or 'CLOSE'.
        """
        output = dict()
        for port in self.targets:
            output[port] = 'CLOSE'
        
        futures = deque()
        with concurrent.futures.ThreadPoolExecutor(max_workers=self.__thread_limit) as executor:
            for port in self.targets:
                future = executor.submit(self.__TCP_connect, ip, port, message)
                futures.append(future)
                while len(futures) >= self.__thread_limit:
                    self.__check_futures(output, futures)
                    time.sleep(0.01)

            # make sure all thread outputs are stored.
            while futures:
                self.__check_futures(output, futures)
                time.sleep(0.01)

        # Print opening ports from small to large
        if self.__verbose:
            for port in self.targets:
                if output[port] == 'OPEN':
                    service = self.__port_map.get(port, None)
                    if service:
                        port_proto = '{}/{}'.format(port, service.proto.upper())
                    else:
                        port_proto = '{}/{}'.format(port, 'UNKNOWN')
                    print('{:10}: {:>10}\n'.format(port_proto, output[port]))

        return output

    @classmethod
    def __check_futures(cls, output, futures):
        """
        Check the executing status of Futures and retrieve the results from them.
        :param output: dict for storing the results
        :param futures: list of concurrent.futures.Future object
        """
        for _ in range(len(futures)):
            future = deque.popleft()
            if future.done():
                try:
                    port, status = future.result()
                    output[port] = status
                except Exception:
                    pass
            else:
                futures.append(future)

    def __TCP_connect(self, ip, port_number, message):
        """
        Perform status checking for a given port on a given ip address using TCP handshake

        :param ip: the ip address that is being scanned
        :type ip: str
        :param port_number: the port that is going to be checked
        :type port_number: int
        :param message: the message that is going to be included in the scanning packets,
        in order to prevent ethical problem, default to ''.
        :type message: str
        """
        # Initialize the TCP socket object based on different operating systems.
        # All systems except for 'Windows' will be treated equally.
        curr_os = platform.system()
        if curr_os == 'Windows':
            TCP_sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
            TCP_sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
            TCP_sock.settimeout(self.__timeout)
        else:
            TCP_sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
            TCP_sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEPORT, 1)
            TCP_sock.settimeout(self.__timeout)

        b_message = message.encode('utf-8', errors='replace')
        # Initialize a UDP socket to send scanning alert message if there exists an non-empty message
        if message:
            UDP_sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
            UDP_sock.sendto(b_message, (ip, int(port_number)))

        try:
            result = TCP_sock.connect_ex((ip, int(port_number)))
            if message != '':
                TCP_sock.sendall(b_message)

            TCP_sock.close()
            # If the TCP handshake is successful, the port is OPEN. Otherwise it is CLOSE
            if result == 0:
                return port_number, 'OPEN'
            else:
                return port_number, 'CLOSE'

        except socket.error as e:
            # Failed to perform a TCP handshake means the port is probably close.
            return port_number, 'CLOSE'
